k2/csrc/fsa_utils.cu

```diff
@@ -2096,4 +2097,395 @@ void FixNumStates(FsaVec *fsas) {
+      }
+      arc_inv_tots_data[i] = inv_tot;
+    });
+  K2_EVAL(c, num_arcs, lambda_modify_cdf, (int32_t arc_idx012) {
+      int32_t state_idx01 = fsa_row_ids2[arc_idx012];
+      FloatType cdf_value = arc_cdf_data[arc_idx012],
+          cdf_value_modified = cdf_value * arc_inv_tots_data[state_idx01];
+      arc_cdf_data[arc_idx012] = cdf_value_modified;
+    });
+  return arc_cdf;
+}
+
+
+
+namespace random_paths_internal {
+
+
+// shared state (the changing part of it) for the algorithm that gets the paths.
+template <typename FloatType>
+struct PathState {
+  int32_t begin_arc_idx01x;  // first arc_idx012 leavng this state..
+  int32_t num_arcs;          // num arcs leaving this state.
+
+  // `p` is a number in the interval [0, 1], which you can think of as being
+  // random (although it's initialized deterministically at fixed intervals).
+  // As we advance `cur_state_idx01` from the start state to later states in the
+  // path, we zoom in more and more; and `p` is always the position within the
+  // probability interval spanned by arcs leaving `cur_state_idx01`.
+  FloatType p;
+};
+
+}  // namespace random_path_internal
+
+template <typename FloatType>
+Ragged<int32_t> RandomPaths(Ragged<FsaVec> &fsas,
+                            Array1<FloatType> &arc_cdf,
+                            const Array1<int32_t> &num_paths,
+                            Ragged<int32_t> &state_batches) {
+  using namespace random_path_internal;
+  K2_CHECK_EQ(fsas.NumAxes(), 3);
+  K2_CHECK_EQ(fsas.NumElements(), arc_cdf.Dim());
+  K2_CHECK_EQ(fsas.Dim0(), num_paths.Dim());
+  K2_CHECK(IsCompatible(fsas, arc_cdf));
+  K2_CHECK(IsCompatible(fsas, num_paths));
+  K2_CHECK(IsCompatible(fsas, state_batches));
+
+
+  ContextPtr c = fsas.Context();
+  int32_t num_fsas = fsas.Dim0();
+
+  // For each FSA, work out the number of batches of states that it has,
+  // which will tell us how much memory we need to allocate.
+  Array1<int32_t> num_state_batches(c, num_fsas);
+  num_state_batches = -1;  // so we can more easily detect errors..
+  int32_t max_batches = state_batches.Dim0();
+  const int32_t *state_batches_row_splits1 = state_batches.RowSplits(1),
+      *state_batches_row_splits2 = state_batches.RowSplits(2);
+  int32_t *num_state_batches_data = num_state_batches.Data();
+  K2_EVAL2(c, num_fsas, max_batches, lambda_set_num_state_batches, (int32_t i, int32_t b) {
+      int32_t this_batch_start = state_batches_row_splits1[b],
+         num_states = state_batches_row_splits2[this_batch_start + i + 1] -
+          state_batches_row_splits2[this_batch_start + i];
+      if (num_states == 0) {
+        num_state_batches_data[i] = 0;
+      } else {
+        int32_t prev_batch_start = state_batches_row_splits1[b - 1],
+            prev_num_states = state_batches_row_splits2[prev_batch_start + i + 1] -
+              state_batches_row_splits2[prev_batch_start + i];
+        if (prev_num_states != 0)
+          num_state_batches_data[i] = 0;
```

Comment:
I guess this function has not finished, but just a question, wondering why not we pass `GetStatesBatch(fsas, false)` as `state_batches` here? Is is becasue we called `GetStatesBatch(fsas, true)` for all GetForward/Backward computation? Can we cache another variable like `state_batch_transpose_` in Python code?

Reply:
Mm.  it's true that it would be more efficient.  But let's wait to see whether there's any problem here. 
When we're ready to test things at the python level I'll look at the nsys output.